### DISCUSSION TITLE
Font-family setting not applied correctly when fontconfig overrides monospace

### DISCUSSION THREAD

**Discussion by warbacon** [ghostty-org/ghostty#3015] [Feature Requests, Ideas] [answered]
# Description
When a system-wide font configuration exists in `~/.config/fontconfig/fonts.conf` that maps the `monospace` family to a specific font (in this case, JetBrains Mono), the terminal ignores the `font-family` setting in its configuration file.

My `~/.config/fontconfig/fonts.conf` file:

```xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
  <match target="pattern">
    <test name="family" compare="eq">
      <string>monospace</string>
    </test>
    <edit name="family" mode="assign" binding="strong">
      <string>JetBrains Mono</string>
    </edit>
  </match>
</fontconfig>
```

- The issue persists even after configuration reload.
- It works fine if this configuration is not present.

# Expected Behavior

The terminal's `font-family` configuration should take precedence over system-wide font settings.

# Steps to Reproduce

1. Set up system-wide font configuration as shown above.
2. Configure terminal with a different font family.
3. Launch terminal.
4. Observe that JetBrains Mono is used instead of the configured font.

# Environment

- **OS**: Arch Linux with Hyprland.
- **Build Method**: Compiled using `zig build -p $HOME/.local -Doptimize=ReleaseFast`.

**Comment by mitchellh** [comment]
I want to discuss this a bit before classifying it as a bug. Fontconfigs APIs are notoriously poorly documented so part of this is simply an ask of what I should be doing instead.

Right now we’re using the fc APIs to ask fc to give us the best match for a given family name and we use whatever fc tells us to use given its own internal scoring.

Do you know what the alternative is?

**Reply by heddxh** [reply]
Maybe we should pass the font-family value specified in config file to Fc API instead of only `monospace`?

**Comment by warbacon** [comment]
I apologize for not being very familiar with this subject and, therefore, unable to suggest an alternative. However, what I do know about `fontconfig` is that it allows setting aliases and fallbacks for fonts. Most of my knowledge comes from the [Arch Wiki](https://wiki.archlinux.org/title/Font_configuration) and other online resources.

Actually this is my complete configuration:

```xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
  <!-- Configuration for the sans-serif font family -->
  <match target="pattern">
    <!-- Check if the font family requested is "sans-serif" -->
    <test name="family" compare="eq">
      <string>sans-serif</string>
    </test>
    <!-- Replace the default "sans-serif" font with "Inter Variable" -->
    <edit name="family" mode="assign" binding="strong">
      <string>Inter Variable</string>
    </edit>
    <!-- Append "Noto Sans" as a fallback font for "sans-serif" -->
    <edit name="family" mode="append" binding="strong">
      <string>Noto Sans</string>
    </edit>
  </match>

  <!-- Configuration for the monospace font family -->
  <match target="pattern">
    <!-- Check if the font family requested is "monospace" -->
    <test name="family" compare="eq">
      <string>monospace</string>
    </test>
    <!-- Replace the default "monospace" font with "JetBrains Mono" -->
    <edit name="family" mode="assign" binding="strong">
      <string>JetBrains Mono</string>
    </edit>
    <!-- Append fallback fonts for "monospace":
         - "Noto Sans Mono" as a general-purpose monospace font
         - "Symbols Nerd Font" for symbols or icon-like characters -->
    <edit name="family" mode="append" binding="strong">
      <string>Noto Sans Mono</string>
      <string>Symbols Nerd Font</string>
    </edit>
  </match>
</fontconfig>
```

With the monospace configuration, I ensure that most applications requesting a monospace font use JetBrains Mono. Additionally, the Nerd Fonts icons are displayed correctly.

This setup has worked well for me across multiple terminals, including Kitty, WezTerm, and Alacritty. Applications generally respect the specified font preferences.

I hope that some of what I've shared here is helpful to you, even if only to a small extent. Please feel free to reach out if you need further assistance—I am more than willing to help! 🫡

**Comment by Promix953** [comment]
Temporary fix:

``` xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
......
  <!-- Ghostty -->
  <match>
    <test qual="all" name="prgname" target="pattern" compare="eq">
      <string>ghostty</string>
    </test>
    <test qual="any" name="family">
      <string>monospace</string>
    </test>
    <edit name="family" mode="prepend" binding="strong">
      <string>My Favorite Font</string>
      <string>Fallback Font</string>
    </edit>
  </match>
......
</fontconfig>
```

**Reply by warbacon** [reply]
I didn't know you could check the program, thanks!

Your fix gave me the idea to make my rule apply only if I'm not in ghostty:

```xml
<!-- Configuration for the monospace font family -->
<match target="pattern">
  <!-- Check if the program is NOT ghostty -->
  <test name="prgname" compare="not_eq">
    <string>ghostty</string>
  </test>
  <!-- Check if the font family requested is "monospace" -->
  <test name="family" compare="eq">
    <string>monospace</string>
  </test>
  <!-- Replace the default "monospace" font with "JetBrains Mono" -->
  <edit name="family" mode="assign" binding="strong">
    <string>JetBrains Mono</string>
  </edit>
  <!-- Append fallback fonts for "monospace":
    - "Noto Sans Mono" as a general-purpose monospace font
    - "Symbols Nerd Font" for symbols and icons -->
  <edit name="family" mode="append" binding="strong">
    <string>Noto Sans Mono</string>
    <string>Symbols Nerd Font</string>
  </edit>
</match>
```

**Reply by Promix953** [reply]
Yes, it's a better rule.
But `<test name="application" compare="not_eq">` should be `<test name="prgname" compare="not_eq">`

**Reply by volodymyr-xn** [reply]
Have the same issue. Would be nice if ghostty `font-family` option had precedence over system font.
Config precedence for font-family works in other terminals i use: `kitty`, `alacritty`, `gnome-terminal`

**Reply by YangtseSu** [reply]
Have the same issue.

**Comment by 00-kat** [comment, marked as answer]
Remove the `binding="strong"`; Ghostty is the only app which has issues for this but no one knows why.

For the record, previous posts on the [Ghostty Discord]:
- https://discord.com/channels/1005603569187160125/1295629990406393888
- https://discord.com/channels/1005603569187160125/1321976159130353674
- https://discord.com/channels/1005603569187160125/1322900882572181573

[Ghostty Discord]: https://discord.gg/ghostty

**Reply by YangtseSu** [reply]
To view these links, one needs to register a discord account.

**Reply by 00-kat** [reply]
That is indeed correct.  The issue was that `binding="strong"` needs to be removed, those links have the exact same resolution in them and were simply included for the record.

**Reply by 00-kat** [reply]
Here's a diff of OP's config in case that helps make it more clear.
```diff
--- a.xml	2025-02-09 22:09:54.845443392 +1100
+++ b.xml	2025-02-09 22:10:26.145811591 +1100
@@ -5,7 +5,7 @@
     <test name="family" compare="eq">
       <string>monospace</string>
     </test>
-    <edit name="family" mode="assign" binding="strong">
+    <edit name="family" mode="assign">
       <string>JetBrains Mono</string>
     </edit>
   </match>
```

**Comment by djneades** [comment]
Please also see https://github.com/basecamp/omarchy/discussions/4818#discussioncomment-16529613.
